### PR TITLE
Created a unified Loader class for validation files

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
Due to the extensive use of Validation on files, the `Dependency` class is becoming unyielding. An alternative to this is to use a unified `ValidatorLoader` that simplifies how a file can be loaded and auto-magically validates the file.

The loader has been built so that it can incorporate `save` methods as well as it logical for these to be located with the `load` methods. The `load` and `save` methods are handled by different classes as the steps involved are different and hard to generalise due to validation. However they both inherit from a shared base class that contains the path information.

The syntax for using the `Loader` is as follows:
```
ValidatorLoader.<load|save(data)>.<file_name_method(*additional_parameters)>
```

The file loader can be created from the `config` and exists natively inside the `BaseCommand`.

Also updated the `Commands` class to it dynamically requires all `commands` in the `src/commads` directory. This should make it easier to add and remove commands in the future.

### CODE REVIEW

**Comment by WilliamMcCumstie** [comment]
NOTE: Codeclimate is failing due to the `TODO's` that are in `Validator::Answer`. However fixing all the required changes is to big of a task for a single PR in will be fixed in the future

File: src/validator/answer.rb [hunk 1]
```diff
@@ -30,9 +30,10 @@ module Validator
     class Answer
       ERROR_FILE = File.join(File.dirname(__FILE__), 'errors.yaml').freeze
 
-      def initialize(metalware_config, answer_file)
+      def initialize(metalware_config, answer_file, loader: nil)
```
Comment (WilliamMcCumstie) [inline src/validator/answer.rb]: This was added before the `Loader` was in the `config`. We clean that up when I clean the class up as a whole

Comment (bobwhitelock) [inline src/validator/answer.rb]: OK, sorry I must have missed that :slightly_smiling_face:

**Comment by WilliamMcCumstie** [comment]
Ok I have done a bit more work on the `Loader` and `Saver` and simplified the syntax. The multiple class have been separated into individual files and the purpose has been refined. I have removed a lot of the meta-programming to make it easier. 

I have now created a `FilePath` class to simplify getting the paths to all the files. It is for the most part and extension of `<section>_answer_file` methods in `Config`. However I feel that these methods would be better served in a dedicated class. `FilePath` does have some meta-programming in which it can access the paths stored in `Constants`. It does make the assumption that any constant ending in `_PATH` is indeed a path.

I have also made all the `TODO` changes to `Answers` and reflected the change in `Dependency`

**Comment by WilliamMcCumstie** [comment]
tl;dr; Everything has changed and you will need to review it again

File: src/validator/saver.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# For more information on the Alces Metalware, please visit:
+# https://github.com/alces-software/metalware
+#==============================================================================
+
+require 'file_path'
+require 'data'
+
+module Metalware
+  module Validator
+    class Saver
```
Comment (bobwhitelock) [inline src/validator/saver.rb]: Do we use this class anywhere (outside the tests for it)? If not can we remove.

File: src/validator/loader.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+# https://github.com/alces-software/metalware
+#==============================================================================
+
+require 'file_path'
+require 'validator/answer'
+require 'validator/configure'
+require 'data'
+
+module Metalware
+  module Validator
```
Comment (bobwhitelock) [inline src/validator/loader.rb]: As mentioned in previous review, I think `Validation` might be a better name for this module.

File: src/validator/loader.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    class Loader
+      def initialize(metalware_config)
+        @config = metalware_config
+        @path = config.file_path
+      end
+
+      def configure
```
Comment (bobwhitelock) [inline src/validator/loader.rb]: `configure_data` might be a better name, calling this `configure` implies to me that this will do something like configure the `Loader` instance. You could suffix all these methods with `_data` if you want to be consistent.

File: src/validator/loader.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        Validator::Configure.new(path.configure_file).load
+      end
+
+      def groups_cache
+        Data.load(path.groups_cache)
```
Comment (bobwhitelock) [inline src/validator/loader.rb]: Don't think this method needs to exist, we don't use it outside of tests (afaics) and its existence on the `Validator::Loader` class implies it will validate the loaded file, which it doesn't do.

File: src/file_path.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      config.domain_answers_file
+    end
+
+    def group_answers(group)
+      config.group_answers_file(group)
+    end
+
+    def node_answers(node)
+      config.node_answers_file(node)
+    end
```
Comment (bobwhitelock) [inline src/file_path.rb]: If you give these methods the same name here and in the `Config` class, then you can simplify all these methods to just `delegate` to the `config`.

File: src/command_helpers/base_command.rb [hunk 7]
```diff
@@ -75,6 +75,10 @@ def enforce_dependency
         Dependency.new(config, command_name, dependency_hash).enforce
       end
 
+      def file_loader
```
Comment (WilliamMcCumstie) [inline src/command_helpers/base_command.rb]: It gets used in #147, will look into the error

File: src/config.rb [hunk 8]
```diff
@@ -63,11 +67,26 @@ def initialize(file = nil, options = {})
       end
     end
 
+    def loader
+      @loader ||= Validator::Loader.new(self)
+    end
+
+    def saver
+      @saver ||= Validator::Saver.new(self)
+    end
```
Comment (WilliamMcCumstie) [inline src/config.rb]: Ok, I think at this point I only use the file loader built into the `BaseCommand` so it can be removed from the `Config`

File: src/validator/saver.rb (hunk 2 above)
Comment (WilliamMcCumstie) [inline src/validator/saver.rb]: It will get used in #147

File: src/validator/loader.rb (hunk 3 above)
Comment (WilliamMcCumstie) [inline src/validator/loader.rb]: Will make a trello card for it

File: src/file_path.rb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+    private
+
+    attr_reader :config
+
+    def define_constant_paths
```
Comment (WilliamMcCumstie) [inline src/file_path.rb]: It is used during `initialize` currently it gets indirectly tested. Can make it more explicit

File: src/validator/loader.rb (hunk 5 above)
Comment (WilliamMcCumstie) [inline src/validator/loader.rb]: This file maybe validated in the future however, will talk about it tomorrow

File: src/command_helpers/base_command.rb (hunk 7 above)
Comment (WilliamMcCumstie) [inline src/command_helpers/base_command.rb]: Error fixed

File: src/file_path.rb (hunk 6 above)
Comment (WilliamMcCumstie) [inline src/file_path.rb]: I don't think these methods should be defined in the `config`. At the moment we have paths that can be defined in `Config` and `Constants` which makes it hard to locate where a file is going to be. Also  `answer` files should always be loaded through the validator. So I think the better option is to remove the methods from `Config` and add them to `FilePath`.

File: src/validator/loader.rb (hunk 4 above)
Comment (WilliamMcCumstie) [inline src/validator/loader.rb]: Will make a card for it. Will be easier to make the change after #147 is merged

File: src/validator/loader.rb (hunk 5 above)
Comment (bobwhitelock) [inline src/validator/loader.rb]: If we validate it in the future we can always add it here then, no need to add extra indirection I think :slightly_smiling_face:
